src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,480 @@
+            {
+                return assembly.GetTypes();
+            }
+            catch (ReflectionTypeLoadException e)
+            {
+                Console.WriteLine($"Failed to load all types from {assembly.FullName}. Will proceed with types loaded successfully.");
+                return e.Types.Where(t => t != null);
+            }
+        }
+
+        private static ImmutableArray<DiagnosticAnalyzer> GetAllAnalyzers(IEnumerable<Assembly> assemblies)
+        {
+            var analyzers = ImmutableArray.CreateBuilder<DiagnosticAnalyzer>();
+            foreach (var diagnosticAnalyzer in GetAllSubclasses<DiagnosticAnalyzer>(assemblies))
+            {
+                analyzers.Add(diagnosticAnalyzer);
+            }
+
+            return analyzers.ToImmutable();
+        }
+
+        private static ImmutableDictionary<string, ImmutableList<CodeFixProvider>> GetAllCodeFixers(IEnumerable<Assembly> assemblies)
+        {
+            Dictionary<string, ImmutableList<CodeFixProvider>> providers = new Dictionary<string, ImmutableList<CodeFixProvider>>();
+            foreach (var codeFixProvider in GetAllSubclasses<CodeFixProvider>(assemblies))
+            {
+                foreach (var diagnosticId in codeFixProvider.FixableDiagnosticIds)
+                {
+                    providers.AddToInnerList(diagnosticId, codeFixProvider);
+                }
+            }
+
+            return providers.ToImmutableDictionary();
+        }
+
+        private static IEnumerable<T> GetAllSubclasses<T>(IEnumerable<Assembly> assemblies)
+        {
+            var returnType = typeof(T);
+            foreach (var assembly in assemblies)
+            {
+                foreach (var type in GetLoadableTypes(assembly))
+                {
+                    if (type.IsSubclassOf(returnType) && !type.IsAbstract)
+                    {
+                        yield return (T)Activator.CreateInstance(type);
+                    }
+                }
+            }
+        }
+
+        private static async Task<ImmutableDictionary<ProjectId, ImmutableArray<Diagnostic>>> GetAnalyzerDiagnosticsAsync(Solution solution, string solutionPath, ImmutableArray<DiagnosticAnalyzer> analyzers, Options options, TelemetryCollector telemetryCollector, CancellationToken cancellationToken)
+        {
+            List<KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>> projectDiagnosticTasks = new List<KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>>();
+
+            // Make sure we analyze the projects in parallel
+            foreach (var project in solution.Projects)
+            {
+                if (project.Language != LanguageNames.CSharp)
+                {
+                    continue;
+                }
+
+                projectDiagnosticTasks.Add(new KeyValuePair<ProjectId, Task<ImmutableArray<Diagnostic>>>(project.Id, GetProjectAnalyzerDiagnosticsAsync(analyzers, project, options, telemetryCollector, cancellationToken)));
```

Comment:
I thought we decide not to run things in concurrent for now to make it easier to investigate.

Reply:
💭  It runs sequentially by default, but allows for concurrent runs (at least this should be the case for StyleCopTester).